### ISSUE TITLE
Segmentation fault in shared library

### ISSUE THREAD

**Issue by purgeme** [google-ai-edge/mediapipe#3109]
So I have been trying to make a shared library for a custom project.
I have tried making a shared library with a print "Hello World!" function and that works but as soon as I import anything related to opencv from mediapipe like ( without adding any actual code ):
```
#include "mediapipe/framework/port/opencv_highgui_inc.h"
#include "mediapipe/framework/port/opencv_imgproc_inc.h"
#include "mediapipe/framework/formats/image_frame_opencv.h"
#include "mediapipe/framework/port/opencv_video_inc.h"
```
( These are the exact lines I tested with, even just one can cause crash )
the shared library causes the program to crash with a segmentation fault.
I have used gdb and this was the output when running:
```
(gdb) run
Starting program: /home/mediapipe/Documents/mediapipe/mediapipe/examples/desktop/test2/a.out 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".

Program received signal SIGSEGV, Segmentation fault.
0x00007fffed0115c4 in google::protobuf::internal::AddDescriptors(google::protobuf::internal::DescriptorTable const*) () from /lib/x86_64-linux-gnu/libprotobuf.so.23
(gdb) 
```
It works just fine if I import anything else.
Just leaving them as dependencies in the BUILD file without including them in the program also causes the same issue ( Sorry I am new to bazel and mediapipe ).

Last commit which my local repository is on:
`commit e6c19885c6d3c6f410c730952aeed2852790d306 (tag: v0.8.9, origin/staging, origin/master, origin/HEAD, master)`

**Comment by sureshdagooglecom** [comment]
Hi @Purgeme , did you get a chance to check [this](https://github.com/google/mediapipe/issues/1734) issue.

**Comment by Purgeme** [comment]
I just looked at it and though I could not see how it is related, I did try building after switching to an older version of mediapipe.
I downloaded v8.4 but it is having issues during build as bazel is trying to download rules_cc-master but what is available is rules_cc-main ( probably due to names of branches having been changed ).

Could you explain if switching back to an older version will help or not ?
- If yes, how can I fix this rules_cc-master issue ?
- If not, what would be the course of action to resolve that segmentation fault ? ( I am working in c++ ).

**Comment by mcclanahoochie** [comment]
Are you using Bazel for your entire project? or are you using a different build system ?

Mediapipe has many example of building a cc_library using opencv. Example https://github.com/google/mediapipe/blob/master/mediapipe/framework/formats/BUILD#L180 builds image_frame_opencv.so

**Comment by Purgeme** [comment]
Sorry for the delay, I am not using Bazel for the entire project. Just for exporting a shared library.

**Comment by sureshdagooglecom** [comment]
@Purgeme ,
did you get required information ?

**Comment by dudutudou** [comment]
Hi @purgeme Did you solve this probleam?

**Comment by purgeme** [comment]
Sorry, been busy with college.
I retried creating a dll and everything works fine now. I am able to create a .so file and even use holistic tracking inside another cpp program with help of the .so file. I tried just creating a program that includes these files and does nothing, still compiles and works.
( It isn't fully self-contained as the graphs still require tflite files at specific relative paths )

Basically, the issue is fixed.
Thank you.

**Comment by Darek07** [comment]
If you comment libopencv_video.so in third_party/opencv_linux.BUILD the problem will dissapear:
`linkopts = [
        "-l:libopencv_core.so",
        "-l:libopencv_calib3d.so",
        "-l:libopencv_features2d.so",
        "-l:libopencv_highgui.so",
        "-l:libopencv_imgcodecs.so",
        "-l:libopencv_imgproc.so",
        #"-l:libopencv_video.so",
        "-l:libopencv_videoio.so",
    ],`
    
    But in this case there is another problem. The line below throws error in demo_run_graph_main:
    `MP_RETURN_IF_ERROR(graph.AddPacketToInputStream(
        inputStream, Adopt(imageFrame.release()).At(Timestamp(frameTimestampUs))));`

**Comment by Darek07** [comment]
Thanks a lot. That's helped.

**Comment by jinpengfei** [comment]
Hi,do you get the information you need?

**Comment by jinpengfei** [comment]
I encountered the same issue as you: when compiling and building the library, it used Protobuf version 3.19 from the WORKSPACE, but at runtime it was using the 3.23 version installed on Ubuntu.

**Comment by akashvverma1995** [comment]
Hi @purgeme 

Could you please confirm whether this issue has been resolved or if it is still ongoing? If it has been fixed, feel free to close the issue. Otherwise, please let us know.
